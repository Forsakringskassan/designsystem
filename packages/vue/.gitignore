# vite dev
/src/vite-dev/app.vue
