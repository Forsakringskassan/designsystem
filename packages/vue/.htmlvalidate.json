{
    "extends": ["@fkui/vue:recommended"],
    "plugins": ["./htmlvalidate"],
    "transform": {
        "^.*\\.spec.ts$": "html-validate-vue:html"
    }
}
