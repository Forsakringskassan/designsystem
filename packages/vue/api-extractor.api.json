{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/api.d.ts",
    "dtsRollup": {
        "enabled": false
    },
    "messages": {
        "compilerMessageReporting": {
            "default": {
                "logLevel": "none"
            }
        }
    }
}
