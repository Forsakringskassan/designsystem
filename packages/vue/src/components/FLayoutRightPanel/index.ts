export { default as FLayoutRightPanel } from "./FLayoutRightPanel.vue";
export {
    type FLayoutRightPanelContentEvent,
    type FLayoutRightPanelInteface,
    FLayoutRightPanelService,
} from "./services";
