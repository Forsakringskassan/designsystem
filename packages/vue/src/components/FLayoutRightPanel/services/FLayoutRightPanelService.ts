import { EventBus } from "../../../utils/event-bus";

/**
 * @public
 */
export type FLayoutRightPanelContentEvent = "close" | "open-dialog" | "open";

/**
 * @public
 */
export interface FLayoutRightPanelInteface {
    open(): void;
    openDialog(title: string): void;
    close(): void;
    on(event: FLayoutRightPanelContentEvent, callback: () => void): void;
    off(event: FLayoutRightPanelContentEvent, callback: () => void): void;
}

/**
 * This service used to open and close the secondary panel.
 * The service has on/off methods to listen to these events
 * that is used internally by the layout components.
 *
 * @internal
 */
class FRightPanelServiceImpl implements FLayoutRightPanelInteface {
    private focusedElementBeforeOpenining: Element | null = null;

    public open(): void {
        this.focusedElementBeforeOpenining = window.document.activeElement;
        this.emit("open");
    }

    public openDialog(title: string): void {
        this.focusedElementBeforeOpenining = window.document.activeElement;
        this.emit("open-dialog", title);
    }

    public close(): void {
        this.emit("close");
        if (this.focusedElementBeforeOpenining) {
            (this.focusedElementBeforeOpenining as HTMLElement).focus();
        }
    }

    public on(
        event: FLayoutRightPanelContentEvent,
        callback: () => void,
    ): void {
        EventBus.$on(`application-layout:${event}`, callback);
    }

    public off(
        event: FLayoutRightPanelContentEvent,
        callback: () => void,
    ): void {
        EventBus.$off(`application-layout:${event}`, callback);
    }

    private emit(event: FLayoutRightPanelContentEvent, arg?: unknown): void {
        EventBus.$emit(`application-layout:${event}`, arg);
    }
}

/**
 * @public
 */
export const FLayoutRightPanelService: FLayoutRightPanelInteface =
    /* @__PURE__ */
    new FRightPanelServiceImpl();
