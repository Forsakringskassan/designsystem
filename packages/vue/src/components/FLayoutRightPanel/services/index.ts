export {
    type FLayoutRightPanelInteface,
    type FLayoutRightPanelContentEvent,
    FLayoutRightPanelService,
} from "./FLayoutRightPanelService";
