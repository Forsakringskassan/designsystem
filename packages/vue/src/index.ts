export * from "./components";
export {
    config,
    type FKUIConfig,
    FKUIConfigButtonOrder,
    type FKUIContext,
    type MaybeWithFKUIContext,
    setRunningContext,
} from "./config";
export * from "./internal-components";
export * from "./plugins";
export * from "./types";
export * from "./utils";
