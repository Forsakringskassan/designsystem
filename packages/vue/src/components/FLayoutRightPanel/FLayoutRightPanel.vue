<script lang="ts">
import { defineComponent, toRefs } from "vue";
import { focus } from "@fkui/logic";
import { getElementFromVueRef } from "../../utils";
import { FIcon } from "../FIcon";
import { useLayoutPanel } from "../FLayoutLeftPanel";
import { FLayoutRightPanelService } from "./services";

export default defineComponent({
    name: "FLayoutRightPanel",
    components: {
        FIcon,
    },
    props: {
        /**
         * The default width for the panel in pixels
         */
        initialWidth: {
            type: String,
            default: "320",
        },
        /**
         * The default maximun width for the panel in percentage where 1 is 100% and 0 is 0%
         */
        maxWidth: {
            type: Number,
            default: 0.5,
            validator(value: unknown): boolean {
                return typeof value === "number" && value > 0 && value <= 1;
            },
        },
        /**
         * The default minium width for the panel in pixels
         */
        minWidth: {
            type: String,
            default: "150",
        },
    },
    setup(props) {
        const { initialWidth, minWidth, maxWidth } = toRefs(props);
        return useLayoutPanel({
            initialWidth,
            minWidth,
            maxWidth,
            grow: "left",
        });
    },
    data() {
        return {
            isAbsolutePositioned: false,
            isOpen: false,
            offsetTop: 0,
        };
    },
    computed: {
        secondaryStyle(): Record<string, string> {
            if (this.isOpen) {
                return { width: `${this.panelWidth}px`, top: `${this.offsetTop}px` };
            }
            return { top: `${this.offsetTop}px` };
        },
        primaryStyle(): Record<string, string> {
            if (this.isOpen && !this.isAbsolutePositioned) {
                return { "margin-right": `${this.panelWidth}px` };
            }
            return {};
        },
        // This is to make word-wrap work in IE11
        contentStyle(): Record<string, string> {
            return { "max-width": `${this.panelWidth - 35}px` };
        },
    },
    mounted(): void {
        const headers = document.getElementsByClassName("layout-application-template__header");
        const header = headers.item(0) as HTMLElement;
        if (header) {
            this.offsetTop = header.offsetHeight;
        }
        FLayoutRightPanelService.on("open", this.onOpenSecondary);
        FLayoutRightPanelService.on("close", this.onCloseSecondary);
        window.addEventListener("resize", this.onResize);
        this.onResize();
    },
    unmounted(): void {
        FLayoutRightPanelService.off("open", this.onOpenSecondary);
        FLayoutRightPanelService.off("close", this.onCloseSecondary);
    },
    methods: {
        async onOpenSecondary(): Promise<void> {
            this.isOpen = true;
            await this.$nextTick();
            const element = getElementFromVueRef(this.$refs["title"]);
            const heading = element.querySelector("h1, h2, h3, h4, h5, h6");
            focus(heading, { force: true });
        },
        onCloseSecondary(): void {
            this.isOpen = false;
        },
        openSecondary(): void {
            FLayoutRightPanelService.open();
        },
        onClickCloseSecondary(): void {
            FLayoutRightPanelService.close();
        },
        onResize(): void {
            this.isAbsolutePositioned = window.innerWidth < 1280;
        },
    },
});
</script>

<template>
    <div class="layout-secondary">
        <!-- [html-validate-disable-next no-inline-style] -->
        <div
            id="layout-secondary__primary"
            class="layout-secondary__primary"
            :class="rightPrimaryClasses"
            :style="primaryStyle"
        >
            <!--
@slot Slot for displaying the primary content.
            -->
            <slot name="default"></slot>
        </div>
        <!-- [html-validate-disable-next no-inline-style] -->
        <aside v-if="isOpen" :style="secondaryStyle" class="layout-secondary__secondary">
            <div class="layout-secondary__secondary__border" @mousedown="onBorderMouseDown">
                <div class="layout-secondary__secondary__border__dot"></div>
                <div class="layout-secondary__secondary__border__dot"></div>
                <div class="layout-secondary__secondary__border__dot"></div>
            </div>
            <div class="layout-secondary__secondary__inner">
                <!-- [html-validate-disable-next no-inline-style] -->
                <div ref="title" class="layout-secondary__secondary__inner__title" :style="contentStyle">
                    <!-- @slot Slot for heading -->
                    <slot name="heading"></slot>
                </div>
                <div>
                    <hr aria-hidden="true" />
                </div>
                <!-- [html-validate-disable-next no-inline-style] -->
                <div
                    class="layout-secondary__secondary__inner__content scroll-target"
                    :class="gridClasses"
                    :style="contentStyle"
                >
                    <!--
@slot Slot for displaying content in the secondary-panel
                    -->
                    <slot name="content"></slot>
                    <button
                        class="button button--tertiary button--small button--tertiary--black layout-secondary__secondary__inner__close"
                        type="button"
                        @click="onClickCloseSecondary"
                    >
                        <span class="sr-only">Stäng sekundärpanelen</span>
                        <f-icon class="button__icon" name="close" />
                    </button>
                </div>
            </div>
        </aside>
    </div>
</template>
