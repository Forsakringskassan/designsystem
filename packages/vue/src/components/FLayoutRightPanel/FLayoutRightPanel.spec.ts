import "html-validate/jest";
import "@fkui/test-utils/jest";
import { flushPromises, mount, VueWrapper } from "@vue/test-utils";
import { createPlaceholderInDocument } from "@fkui/test-utils/vue";
import FLayoutRightPanel from "./FLayoutRightPanel.vue";
import { FLayoutRightPanelService } from "./services/FLayoutRightPanelService";

let wrapper: VueWrapper;

const defaultSlots = {
    default: "DEFAULT",
    content: "TOPNAVIGATION",
    heading: /* HTML */ ` <h3>TITEL</h3> `,
};

async function createWrapper(): Promise<VueWrapper> {
    const wrapper = mount(FLayoutRightPanel, {
        slots: {
            ...defaultSlots,
        },
        global: {
            stubs: ["FIcon"],
        },
        attachTo: createPlaceholderInDocument(),
    });
    FLayoutRightPanelService.open();
    // wait for it to open
    await wrapper.vm.$nextTick();
    await flushPromises();
    return wrapper;
}

describe("snapshot", () => {
    it("should match snapshot", async () => {
        wrapper = await createWrapper();
        expect(wrapper).toMatchSnapshot();
    });
});

it("should be closable", async () => {
    wrapper = await createWrapper();
    await wrapper.get("button").trigger("click");
    await flushPromises();
    expect(wrapper.find(".layout-secondary__secondary").exists()).toBeFalsy();
});

it("should focus title on open", async () => {
    wrapper = await createWrapper();
    const heading = wrapper.get("h3");
    expect(heading.element).toHaveFocus();
});

describe("html-validate", () => {
    it("should allow defined slots", () => {
        expect.assertions(1);
        const slotTemplates = Object.entries(defaultSlots).map(
            ([key, value]) => `<template #${key}>${value}</template>`,
        );

        expect(
            `<f-layout-right-panel>${slotTemplates}</f-layout-right-panel>`,
        ).toHTMLValidate();
    });
});
