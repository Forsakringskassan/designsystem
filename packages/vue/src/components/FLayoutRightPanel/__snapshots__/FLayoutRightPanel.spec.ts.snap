// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot should match snapshot 1`] = `
<div class="layout-secondary">
  <div id="layout-secondary__primary" class="layout-secondary__primary">DEFAULT</div>
  <aside style="width: 320px; top: 0px;" class="layout-secondary__secondary">
    <div class="layout-secondary__secondary__border">
      <div class="layout-secondary__secondary__border__dot"></div>
      <div class="layout-secondary__secondary__border__dot"></div>
      <div class="layout-secondary__secondary__border__dot"></div>
    </div>
    <div class="layout-secondary__secondary__inner">
      <div class="layout-secondary__secondary__inner__title" style="max-width: 285px;">
        <h3 tabindex="-1">TITEL</h3>
      </div>
      <div>
        <hr aria-hidden="true">
      </div>
      <div class="layout-secondary__secondary__inner__content scroll-target" style="max-width: 285px;">TOPNAVIGATION<button class="button button--tertiary button--small button--tertiary--black layout-secondary__secondary__inner__close" type="button"><span class="sr-only">Stäng sekundärpanelen</span>
          <f-icon-stub name="close" library="f" class="button__icon"></f-icon-stub>
        </button></div>
    </div>
  </aside>
</div>
`;
